File: src/assets/customJs/center/statistics.js
document.addEventListener("DOMContentLoaded", () => {
  const apiEndpoint = "/api/statistics/";
  const totalBranches = document.getElementById("totalBranches");
  const totalProfessions = document.getElementById("totalProfessions");
  const totalCourses = document.getElementById("totalCourses");
  const totalGroups = document.getElementById("totalGroups");

  // List Containers
  const branchLocations = document.getElementById("branchLocations");
  const professionFieldList = document.getElementById("professionFieldList");
  const activeGroupsList = document.getElementById("activeGroupsList");

  // Charts
  const branchGrowthChartCtx = document.getElementById("branchGrowthChart").getContext("2d");
  const professionDistributionChartCtx = document.getElementById("professionDistributionChart").getContext("2d");
  const weeklyScheduleChartCtx = document.getElementById("weeklyScheduleChart").getContext("2d");
  const courseRevenueChartCtx = document.getElementById("courseRevenueChart").getContext("2d");

  let branchGrowthChart, professionDistributionChart, weeklyScheduleChart, courseRevenueChart;

  // Fetch data from API
  fetch(apiEndpoint)
    .then((response) => response.json())
    .then((data) => {
      if (data.success) {
        const stats = data.data;

        // Update key statistics
        totalBranches.textContent = stats.branches.total;
        totalProfessions.textContent = stats.professions.total;
        totalCourses.textContent = stats.courses.total;
        totalGroups.textContent = stats.groups.total;

        // Populate branch locations
        branchLocations.innerHTML = stats.branches.locations
          .map(
            (branch) =>
              `<li class="list-group-item">
                <strong>${branch.location}</strong> - Contact: ${branch.contact}, Telegram: ${branch.telegram}
              </li>`
          )
          .join("");

        // Populate profession-field list
        professionFieldList.innerHTML = stats.fields.kasb_distribution
          .map(
            (field) =>
              `<li class="list-group-item">
                <strong>${field.kasb__nomi}</strong> - Yo'nalishlar soni: ${field.count}
              </li>`
          )
          .join("");

        // Populate active groups
        activeGroupsList.innerHTML = stats.groups.recent_groups
          .map(
            (group) =>
              `<li class="list-group-item">
                <strong>${group.group_name}</strong> - Kurs: ${group.kurs__nomi}, Faolmi: ${group.is_active ? "Ha" : "Yo'q"}
              </li>`
          )
          .join("");

        // Initialize Branch Growth Chart
        if (branchGrowthChart) branchGrowthChart.destroy();
        branchGrowthChart = new Chart(branchGrowthChartCtx, {
          type: "line",
          data: {
            labels: ["Oxirgi yil", "Oxirgi oy", "Bugun"],
            datasets: [
              {
                label: "Filiallar",
                data: [stats.branches.last_year, stats.branches.last_month, stats.branches.today],
                borderColor: "rgba(75, 192, 192, 1)",
                fill: false,
              },
            ],
          },
        });

        // Initialize Profession Distribution Chart
        if (professionDistributionChart) professionDistributionChart.destroy();
        professionDistributionChart = new Chart(professionDistributionChartCtx, {
          type: "pie",
          data: {
            labels: stats.fields.kasb_distribution.map((item) => item.kasb__nomi),
            datasets: [
              {
                data: stats.fields.kasb_distribution.map((item) => item.count),
                backgroundColor: ["#4caf50", "#ff9800", "#f44336", "#2196f3", "#9c27b0"],
              },
            ],
          },
        });

        // Initialize Weekly Schedule Chart
        if (weeklyScheduleChart) weeklyScheduleChart.destroy();
        weeklyScheduleChart = new Chart(weeklyScheduleChartCtx, {
          type: "bar",
          data: {
            labels: ["Dushanba", "Seshanba", "Chorshanba", "Payshanba", "Juma", "Shanba", "Yakshanba"],
            datasets: [
              {
                label: "Guruhlar soni",
                data: calculateWeeklySchedule(stats.weekly_schedule),
                backgroundColor: "#ff9800",
              },
            ],
          },
        });

        // Initialize Course Revenue Chart
        if (courseRevenueChart) courseRevenueChart.destroy();
        courseRevenueChart = new Chart(courseRevenueChartCtx, {
          type: "doughnut",
          data: {
            labels: stats.courses.recent_courses.map((course) => course.nomi),
            datasets: [
              {
                data: stats.courses.recent_courses.map((course) => course.narxi),
                backgroundColor: ["#f44336", "#4caf50", "#2196f3", "#9c27b0", "#ff9800"],
              },
            ],
          },
        });
      } else {
        toastr.error("Statistikani yuklashda xatolik yuz berdi.", "Xatolik");
      }
    })
    .catch(() => toastr.error("Statistikani yuklashda xatolik yuz berdi.", "Xatolik"));

  // Helper function to calculate weekly schedule counts
  function calculateWeeklySchedule(weeklyData) {
    const weekDays = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"];
    const dayCounts = weekDays.reduce((acc, day) => ({ ...acc, [day]: 0 }), {});

    weeklyData.forEach((group) => {
      group.days_of_week.forEach((day) => {
        if (dayCounts[day] !== undefined) {
          dayCounts[day]++;
        }
      });
    });

    return weekDays.map((day) => dayCounts[day]);
  }
});

File: api/center/statistics.py
from django.http import JsonResponse
from django.views import View
from center.models import Center, Filial, Kasb, Yonalish, Kurs, E_groups, GroupMembership
from django.db.models import Count, Sum, F
from datetime import datetime, timedelta


class StatisticsDashboardView(View):
    def get(self, request):
        today = datetime.now()
        last_month = today - timedelta(days=30)
        last_year = today - timedelta(days=365)

        # Filial Statistics
        filial_stats = {
            "total": Filial.objects.count(),
            "today": Filial.objects.filter(created_at__date=today.date()).count(),
            "last_month": Filial.objects.filter(created_at__gte=last_month).count(),
            "last_year": Filial.objects.filter(created_at__gte=last_year).count(),
            "locations": list(Filial.objects.values("location", "contact", "telegram")),
        }

        # Kasb Statistics
        kasb_stats = {
            "total": Kasb.objects.count(),
            "active": Kasb.objects.filter(is_active=True).count(),
            "recent": list(Kasb.objects.order_by("-created_at")[:5].values("nomi", "created_at")),
        }

        # Yonalish Statistics
        yonalish_stats = {
            "total": Yonalish.objects.count(),
            "active": Yonalish.objects.filter(is_active=True).count(),
            "kasb_distribution": list(
                Yonalish.objects.values("kasb__nomi").annotate(count=Count("id")).order_by("-count")
            ),
        }

        # Kurs Statistics
        kurs_stats = {
            "total": Kurs.objects.count(),
            "total_revenue": Kurs.objects.aggregate(Sum("narxi"))["narxi__sum"] or 0,
            "recent_courses": list(
                Kurs.objects.order_by("-created_at").values("nomi", "narxi", "created_at")[:5]
            ),
        }

        # Group Statistics
        group_stats = {
            "total": E_groups.objects.count(),
            "active_groups": E_groups.objects.filter(is_active=True).count(),
            "students_total": GroupMembership.objects.count(),
            "group_revenue": E_groups.objects.annotate(group_revenue=F("kurs__narxi"))
            .aggregate(total_revenue=Sum("group_revenue"))["total_revenue"]
            or 0,
            "recent_groups": list(
                E_groups.objects.order_by("-created_at").values(
                    "group_name", "kurs__nomi", "created_at", "is_active"
                )[:5]
            ),
        }

        # Student Statistics in Groups
        group_student_distribution = list(
            GroupMembership.objects.values("group__group_name")
            .annotate(student_count=Count("student"))
            .order_by("-student_count")
        )

        # Weekly Class Schedule
        weekly_schedule = list(
            E_groups.objects.values("group_name", "days_of_week", "kurs__nomi")
        )

        # Aggregated Data
        data = {
            "branches": filial_stats,
            "professions": kasb_stats,
            "fields": yonalish_stats,
            "courses": kurs_stats,
            "groups": group_stats,
            "group_student_distribution": group_student_distribution,
            "weekly_schedule": weekly_schedule,
        }

        return JsonResponse({"success": True, "data": data})
